gameobjects/gameobject.h

```diff
@@ -20,40 +21,60 @@ class GameObject
     GameObject() = default;
     virtual ~GameObject() = default;
 
+    virtual void InitalizeComponents() = 0;
     virtual void OnStart();
     virtual void OnUpdate();
     virtual void OnCollisionStart(GameObject& other);
     virtual void OnCollisionStay(GameObject& other);
     virtual void OnCollisionStop(GameObject& other);
 
     int GetId() const;
-    ColliderBox* GetColliderBox() const;
-    glm::vec3 GetPosition() const;
-    void SetPosition(const glm::vec3& position);
     std::string GetInstanceName() const;
     void SetInstanceName(const std::string& name);
 
-    bool CheckForCollision(GameObject& other);
     void SetTimeout(std::chrono::duration<double> timeout, std::function<void()> callback);
     void PlaySound(const Sound& sound);
     void PlaySound(const Sound& sound, const glm::vec3& position);
 
-    void Render() const;
     template <typename T>
     T* GetObject()
     {
         return dynamic_cast<T*>(this);
     }
 
-protected:
-    std::unique_ptr<Mesh>mMesh { nullptr };
-    std::unique_ptr<ColliderBox>mColliderBox { nullptr };
+    template<typename T>
+    T* GetComponent()
+    {
+        const int componentID = GetComponentTypeID<T>();
+
+        auto iterator = mComponents.find(componentID);
+        if (iterator != mComponents.end())
+        {
+            return static_cast<T*>(mComponents[componentID]);
+        }
+
+        return nullptr;
+    }
+
+    template<typename T, typename... Args>
+    T* AddComponent(Args&&... args)
+    {
+        std::unique_ptr<T> component = std::make_unique<T>(std::forward<Args>(args)...);
```

Comment:
now we are modern c++ing. This is an advanced ass statement.

`auto` could be used here, unique_ptr is obvious